File: src/scoring/scoreFifteens.ts
import { Card } from "../interfaces/card.interface";
import { getCombinations } from "../utils/getCombinations";

/**
 * Scores 2 points for each combination of cards that sums to 15.
 */
export const scoreFifteens = (allCards: Card[]): number => {
  let score = 0;
  for (let i = 2; i <= allCards.length; i++) {
    const combinations = getCombinations(allCards, i);
    for (const combo of combinations) {
      if (combo.reduce((acc, card) => acc + card.value, 0) === 15) {
        score += 2;
      }
    }
  }
  return score;
};

File: crib_calc/scoring/score_fifteens.py
from crib_calc.card import Card
from crib_calc.utils.get_combinations import get_combinations


def score_fifteens(all_cards: list[Card]) -> int:
    score = 0
    for size in range(2, len(all_cards) + 1):
        for combo in get_combinations(all_cards, size):
            if sum(c.value for c in combo) == 15:
                score += 2
    return score
